### PR TITLE
command: implement empty /history

### PR DESCRIPTION
**Pull request by aviau**

### CODE REVIEW

**Comment by aviau** [comment]
Exemple de moi qui a écrit ``/history allo``

```
[2020-07-28T21:53:01Z ERROR purple_icq] cmd_func: conv=Crawl Me! cmd=history args=["allo"]
```

**Comment by aviau** [comment]
@isra17 review plz!

I suggest to limit this PR's scope here and resume in the next PR.


Example of me typing ``/history allo``

```
[2020-07-28T21:53:01Z ERROR purple_icq] cmd_func: conv=Crawl Me! cmd=history args=["allo"]
```

And now ``/help history``

Screenshot from 2020-07-28 17-56-48

File: src/lib.rs [hunk 1]
```diff
@@ -398,6 +403,23 @@ impl purple::InputHandler for PurpleICQ {
     }
 }
 
+impl purple::CommandHandler for PurpleICQ {
+    fn command(
+        &mut self,
+        conversation: &mut ChatConversation,
```
Comment (leameow) [inline src/lib.rs:409]: TODO: Switch a Conversation quand je merge mon autre branch #37

File: src/lib.rs [hunk 2]
```diff
@@ -398,6 +403,23 @@ impl purple::InputHandler for PurpleICQ {
...
+        command: &str,
+        args: Vec<&str>,
+    ) -> purple_sys::PurpleCmdRet {
+        log::error!(
+            "cmd_func: conv={} cmd={} args={:?}",
+            conversation.get_title().unwrap_or("unknown"),
+            command,
+            args
+        );
+        purple_sys::PurpleCmdRet::PURPLE_CMD_RET_OK
```
Comment (leameow) [inline src/lib.rs]: J'aime bien reimporter ces enum dans `purple::` comme ca `lib.rs` depend uniquement de `purple`

File: src/purple/handlers/traits.rs [hunk 3]
```diff
@@ -51,6 +52,10 @@ pub trait GetChatNameHandler {
     fn get_chat_name(data: Option<&mut StrHashTable>) -> Option<String>;
 }
 
+pub trait HandleCommandHandler {
+    fn handle_command(&mut self, cmd: &str, args: Vec<&str>) -> purple_sys::PurpleCmdRet;
```
Comment (leameow) [inline src/purple/handlers/traits.rs]: While we give us trouble to convert `String` to `&str`, might as well use a slice here! `args: [&str]`. This way all parameters a reference and no owned value (Such as `Vec`) are passed.

File: src/purple/mod.rs [hunk 4]
```diff
@@ -66,3 +70,56 @@ where
...
+    let cmd = CStr::from_ptr(c_cmd).to_str().unwrap();
+    let args: Vec<String> = FromGlibPtrContainer::from_glib_none(c_args);
+    // TODO: This is a conversation. Not necessairly a ChatConersation.
+    let mut conversation =
+        ChatConversation::from_ptr(conversation_ptr as *mut purple_sys::PurpleConvChat).unwrap();
+
+    closure(
+        &mut conversation,
+        cmd,
+        args.iter().map(std::ops::Deref::deref).collect(),
```
Comment (leameow) [inline src/purple/mod.rs]: you can consume `args` with `into_iter()` if you don't plan to use `args` afterward.

File: src/purple/handlers/traits.rs (hunk 3 above)
Comment (aviau) [reply src/purple/handlers/traits.rs]: Done!

File: src/lib.rs (hunk 2 above)
Comment (aviau) [reply src/lib.rs]: Done!

File: src/lib.rs (hunk 1 above)
Comment (aviau) [reply src/lib.rs:409]: yeah je l'avais noté

File: src/purple/mod.rs (hunk 4 above)
Comment (aviau) [reply src/purple/mod.rs]: Hmm I can't seem to do it.

I tried a couple of things but I always end up getting something similar:

```
123 |         args.into_iter().map(|s| s.as_str()).collect(),
    |                                  -^^^^^^^^^
    |                                  |
    |                                  returns a value referencing data owned by the current function
    |                                  `s` is borrowed here
```

Comment (leameow) [reply src/purple/mod.rs]: Make sense!
Nvm the `into_iter` :)

**Comment by aviau** [comment]
@isra17 Any other comments?

File: src/purple/handlers/traits.rs [hunk 5]
```diff
@@ -51,6 +52,10 @@ pub trait GetChatNameHandler {
     fn get_chat_name(data: Option<&mut StrHashTable>) -> Option<String>;
 }
 
+pub trait HandleCommandHandler {
```
Comment (leameow) [inline src/purple/handlers/traits.rs]: Wait, this trait is not actually used, duplicated with `CommandHandler`

File: src/purple/handlers/traits.rs [hunk 6]
```diff
@@ -88,3 +93,25 @@ pub trait InputHandler {
         unsafe { purple_sys::purple_input_remove(input_handle) != 0 }
     }
 }
+
+pub trait CommandHandler {
+    fn command(
+        &mut self,
+        conversation: &mut ChatConversation,
+        command: &str,
+        args: Vec<&str>,
```
Comment (leameow) [inline src/purple/handlers/traits.rs]: slice would come here `[&str]`

**Pull request merged**